sql-plugin/src/main/java/com/nvidia/spark/rapids/UnsafeRowToColumnarBatchIterator.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids;
+
+import ai.rapids.cudf.ColumnVector;
+import ai.rapids.cudf.DType;
+import ai.rapids.cudf.HostColumnVector;
+import ai.rapids.cudf.HostColumnVectorCore;
+import ai.rapids.cudf.HostMemoryBuffer;
+import ai.rapids.cudf.NvtxColor;
+import ai.rapids.cudf.NvtxRange;
+import ai.rapids.cudf.Table;
+import org.apache.spark.TaskContext;
+import org.apache.spark.sql.catalyst.CudfUnsafeRow;
+import org.apache.spark.sql.catalyst.expressions.Attribute;
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow;
+import org.apache.spark.sql.execution.metric.SQLMetric;
+import org.apache.spark.sql.vectorized.ColumnarBatch;
+import scala.collection.Iterator;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+/**
+ * This class converts UnsafeRow instances to ColumnarBatches on the GPU through the magic of
+ * code generation. This just provides most of the framework a concrete implementation will
+ * be generated based off of the schema.
+ */
+public abstract class UnsafeRowToColumnarBatchIterator implements Iterator<ColumnarBatch> {
+  protected final Iterator<UnsafeRow> input;
+  protected UnsafeRow pending = null;
+  protected final int numRowsEstimate;
+  protected final long dataLength;
+  protected final DType[] rapidsTypes;
+  protected final SQLMetric totalTime;
+  protected final SQLMetric numInputRows;
+  protected final SQLMetric numOutputRows;
+  protected final SQLMetric numOutputBatches;
+
+  protected UnsafeRowToColumnarBatchIterator(
+      Iterator<UnsafeRow> input,
+      Attribute[] schema,
+      CoalesceGoal goal,
+      SQLMetric totalTime,
+      SQLMetric numInputRows,
+      SQLMetric numOutputRows,
+      SQLMetric numOutputBatches) {
```

Comment:
Side comment: I'm wondering if we should make a container class for these common metrics or just pass the metrics map to simplify the call signature.  Not a must-fix for this PR.

Reply:
Ya I am seeing more and more of this and keeping the order right can be painful